lucene/core/src/test/org/apache/lucene/index/TestTermsEnumDeadlock.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.lucene.index;
+
+import org.apache.lucene.util.BytesRef;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.lang.management.ManagementFactory;
+import java.lang.management.RuntimeMXBean;
+import java.nio.file.Paths;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+
+public class TestTermsEnumDeadlock extends Assert {
+  private static final int MAX_TIME_SECONDS = 15;
+
+  @Test
+  public void testDeadlock() throws Exception {
+    for (int i = 0; i < 20; i++) {
+      // Fork a separate JVM to reinitialize classes.
```

Comment:
Could we do this by creating a new class loader instead of a whole new process?

Reply:
Thank you for your review @madrob :D

Hmm... Is there any way to do static initialization multiple times in the same JVM?

As far as I know, static initialization is only executed once per JVM and the concepts of loading and initialization are different.
https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-5.html
Class loading : https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-5.html#jvms-5.3
Initialization : https://docs.oracle.com/javase/specs/jvms/se11/html/jvms-5.html#jvms-5.5

And that TC is meaningful only when the test is run the first time or alone.

Please let me know if I'm wrong!

Reply:
Classloading won't help, because we still need a separate JVM. When JVM loads classes, it ttys in parent classliadee first. If the class is already there it won't load. So we need at least one separate process.

I don't like to try many times each with separate JVM. Maybe only try once (like in the other test with codecs). It may not fail every time, but sometimes test fails.

I am also not sure if we really need a test for this. If we may get a static checker that finds classes that initialize their subclasses in their own static initializer, we can prevent similar cars in future.

Reply:
Thank you for your review @uschindler :D

It was added to ensure working.

But I agree with your comment about test iteration.
So I'll delete the for-loop statement and change it to run only once.

As for the necessity of deadlock test, I'll add TODO comment.
Later, if someone applies to static checker about it, I'll delete it at that time.